File: solutions/java/grains/1/src/main/java/Grains.java
import java.math.BigInteger;

public final class Grains {
    public BigInteger grainsOnSquare(final int square) {
        if (square < 1 || square > 64)
            throw new IllegalArgumentException("square must be between 1 and 64");
        
        if (square <= 2)
            return bigIntOf(square);

        BigInteger previousGrain = grainsOnSquare(square - 1);
        return previousGrain.multiply(bigIntOf(2));
    }

    public BigInteger grainsOnBoard() {
        return grainsOnSquare(64).multiply(bigIntOf(2)).subtract(bigIntOf(1));
    }

    private BigInteger bigIntOf(int n) {
        return BigInteger.valueOf(n);
    }
}

File: solutions/kotlin/grains/1/src/main/kotlin/Grains.kt
import java.math.BigInteger

object Board {
    fun getTotalGrainCount(): BigInteger {
        return getTotalGrainCount(1, BigInteger.ZERO)
    }

    private tailrec fun getTotalGrainCount(number: Int, total: BigInteger): BigInteger =
        when {
            number > 64 -> total
            else -> getTotalGrainCount(number + 1, total + getGrainCountForSquare(BigInteger.ONE, number, 1))
        }

    fun getGrainCountForSquare(square: Int): BigInteger {
        require(square in 1..64) {
            "The square should be inbetween 1 and 100."
        }
        return getGrainCountForSquare(BigInteger.ONE, square, 1)
    }

    private tailrec fun getGrainCountForSquare(grains: BigInteger, maxSquare: Int, count: Int): BigInteger = 
        when {
            count == maxSquare -> grains
            else -> getGrainCountForSquare(grains + grains, maxSquare, count + 1)
        }
}
